### DISCUSSION TITLE
sum types accept interface

### DISCUSSION THREAD

**Discussion by ghost** [vlang/v#11812] [Ideas]

**Comment by Delta456** [comment]
How will this work and what's the need of it?

**Comment by spaceface777** [comment]
@kahsa do you have any examples where this feature would be useful?

**Comment by spaceface777** [comment]
NB: this was disallowed in #6232, which seems to be a "temporary" change until codegen bugs were resolved.

**Comment by ghost** [comment]
I wrote these now and denied and know it is not allowed.

1) `type Focusable = ui.Window | ui.Widget`

2)
```v
for child in window.children {
	// child is Widget but check if it implements OnMouseDown too.
	if child is OnMouseDown {
		window.eventbus.publish(event.on_mouse_down, window, e)
	}
}
```

**Comment by ghost** [comment]
`interfase OnMouseDown{ on_mouse_down(event MouseEvetn) }`

`type ChildWidget = ui.Widget | ui.OnMouseDown`

**Comment by spaceface777** [comment]
hmm, yeah, I see how this (along with interface embedding) could be very useful.

```v
interface Widget {
    // some `widget`y things
    size() (int, int)
}
interface FocusableWidget {
    Widget
    focus() bool
}

type ChildWidget = FocusableWidget | Widget

//////////////////

for child in window.children {
	// `child` must always implement Widget
        w, h := child.size()
        // but...
	if child is FocusableWidget {
                // `child` is Focusable as well, call its `focus` method
                child.focus()
	}
}
```

This would also potentially fix `init_once` always being required in vweb issue.

**Comment by rcqls** [comment]
I have no idea if this can be done but I definitely think that interface should be used simply like this (no need of sum type here):
```v
interface Widget {
    // some `widget`y things
    size() (int, int)
}
interface Focusable {
    focus() bool
}

//////////////////

for child in window.children { // Here window.children is []Widget
	// as `child`  is Widget
        w, h := child.size()
        // Dynamic casting locally to other interface should be awesome! 
	if child is Focusable { // check if child  (as the original Struct conforming to Widget interface) is Focusable i.e. responds to focus() method
                // `child` becomes locally Focusable 
                child.focus()
	}
}
```
But this is maybe only a dream. Even though I have some hope, since  I have in mind  the `type_name()` feature introduced for interface by @spaceface777  that proves that original Struct of an interface can be retrieved.
